module-system/module-implementations/sov-chain-state/src/hooks.rs

```
@@ -0,0 +1,83 @@
+use core::fmt;
+
+use borsh::{BorshDeserialize, BorshSerialize};
+use sov_modules_api::hooks::SlotHooks;
+use sov_modules_api::{Context, Spec};
+use sov_rollup_interface::services::da::SlotData;
+use sov_rollup_interface::zk::ValidityCondition;
+use sov_state::{Storage, WorkingSet};
+use thiserror::Error;
+
+use super::ChainState;
+use crate::{StateTransitionId, TransitionInProgress};
+
+#[derive(Debug, Clone, Error)]
+pub(crate) enum ChainStateError {
+    NoTransitionInProgress,
+}
+
+impl fmt::Display for ChainStateError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let error_msg = match self {
+            Self::NoTransitionInProgress => "No transition in progress",
+        };
+
+        write!(f, "{error_msg}")
+    }
+}
+
+impl<Ctx: Context, Cond: ValidityCondition + Default + BorshDeserialize + BorshSerialize>
+    SlotHooks<Cond> for ChainState<Ctx, Cond>
+{
+    type Context = Ctx;
+
+    fn begin_slot_hook(
+        &self,
+        slot: &impl SlotData<Condition = Cond>,
+        working_set: &mut WorkingSet<<Self::Context as Spec>::Storage>,
+    ) -> anyhow::Result<()> {
+        let curr_height = self.slot_height.get_or_err(working_set)?;
+
+        if curr_height == 0 {
+            // First transition right after the genesis block
+            self.genesis_hash.set(
+                &working_set.backing().get_state_root(&Default::default()),
+                working_set,
+            )
+        } else {
+            let transition: StateTransitionId<Cond> = {
+                let last_transition_in_progress = self
+                    .in_progress_transition
+                    .get(working_set)
+                    .ok_or(ChainStateError::NoTransitionInProgress)?;
```

This should be unreachable, right? If so, we should probably panic since the error is unexpected and unrecoverable (and indicates a bug or malicious prover)